src/wagtail_vector_index/storage/models.py

```diff
@@ -220,6 +234,50 @@ def bulk_from_documents(
             seen_keys.add(key)
             yield objects_by_key[key]
 
+    async def abulk_from_documents(
+        self, documents: Iterable[Document]
+    ) -> AsyncGenerator[models.Model, None, None]:
+        """A copy of `bulk_from_documents`, but async"""
+        # Force evaluate generators to allow value to be reused
+        documents = tuple(documents)
+
+        ids_by_content_type: dict[str, list[str]] = defaultdict(list)
+        for doc in documents:
+            ids_by_content_type[doc.metadata["content_type_id"]].append(
+                doc.metadata["object_id"]
+            )
+
+        # NOTE: (content_type_id, object_id) combo keys are required to
+        # reliably map data from multiple models
+        objects_by_key: dict[tuple[str, str], models.Model] = {}
+        for content_type_id, ids in ids_by_content_type.items():
+            model_class = await self._amodel_class_from_ctid(content_type_id)
+            model_objects = model_class.objects.filter(pk__in=ids)
+            objects_by_key.update(
+                {(content_type_id, str(obj.pk)): obj async for obj in model_objects}
+            )
+
+        seen_keys = set()  # de-dupe as we go
+        for doc in documents:
+            key = (doc.metadata["content_type_id"], doc.metadata["object_id"])
+            if key in seen_keys:
+                continue
+            seen_keys.add(key)
+            yield objects_by_key[key]
+
+    @staticmethod
+    async def _aget_content_type_for_id(id: int) -> ContentType:
```

Comment:
as Django's ContentTypeManager class doesn't have a `aget_for_id` method, I had to copy-paste the code of its `get_for_id` method and adapt it here